core/kernel/msg_param.c

```diff
@@ -31,6 +31,121 @@
 #include <kernel/msg_param.h>
 #include <mm/mobj.h>
 
+/**
+ * msg_param_extract_pages() - extract list of pages from
+ * OPTEE_MSG_ATTR_NONCONTIG buffer.
+ *
+ * @buffer:	pointer to parameters array
+ * @pages:	output array of page addresses
+ * @num_pages:  number of pages in array
+ *
+ * return:
+ *   page count on success
+ *   <0 on error
+ *
+ * You can think that @buffer points to PA of the following structure:
+ *
+ * struct page_data {
+ *   uint64_t pages_array[OPTEE_MSG_NONCONTIG_PAGE_SIZE/sizeof(uint64_t) - 1];
+ *   uint64_t next_page_data;
+ * };
+ *
+ * So, it is like linked list of arrays, where every element of linked list fits
+ * exactly into one 4K page.
+ *
+ * This function extracts data from arrays into one array pointed by @pages
+ *
+ * @buffer points to data shared with NW, so some precautions should be taken.
+ */
+static ssize_t msg_param_extract_pages(paddr_t buffer, paddr_t *pages,
+				       size_t num_pages)
+{
+	size_t cnt;
+	struct mobj *mobj;
+	paddr_t page;
+	uint64_t *va;
+
+	if (buffer & SMALL_PAGE_MASK)
+		return TEE_ERROR_BAD_PARAMETERS;
```

Comment:
To return this value the function should use a `TEE_Result` return type. I'm noting that the returned value isn't really used for anything than check that it matches a precomputed value.